docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,309 @@
+time to 20s in the future, the master might refresh its status every
+10s.
+
+The lease master will provide the following service for read leases:
+
+```proto
+message WatchRequest {
+  optional int32 nodeID;
+}
+
+message WatchResponse {
+  enum State {
+    CURRENT
+    DRAINING
+    DEAD
+  }
+  optional int32 descID;
+  optional int32 version;
+  optional State state;
+}
+
+message AcquireRequest {
+  optional int32 nodeID;
+  optional int32 descID;
+  optional int32 version;
+}
+
+message AcquireResponse {
+  optional message<Error> error;
+  optional int64 expiration;
+}
+
+message ReleaseRequest {
+  optional int32 nodeID;
+  optional int32 descID;
+  optional int32 version;
+  optional bool all;  // Release all leases held by the node.
+}
+
+message ReleaseResponse {
+  optional message<Error> error;
+}
+
+message PublishRequest {
+  optional message<TableDescriptor> desc;
+}
+
+message PublishResponse {
+  optional message<Error> error;
+}
+
+service TableLeaseService {
+  // Watch registers the node for updates to descriptor versions. A
+  // watch should be registered before any other lease operation.
+  rpc Watch(WatchRequest) returns (stream WatchResponse) {}
+  
+  // Acquire acquires a read lease for the specified descriptor and
+  // version. The lease is associated with the specified node. An
+  // error will be returned if the specified version is not the most
+  // recent version for the descriptor.
+  rpc Acquire(AcquireRequest) returns (AcquireResponse) {}
+
+  // Release allows a read lease to be released before it
+  // expires. This should be used when a watch request indicates a new
+  // version of a descriptor is available and the lease holder
+  // determines it is no longer using the old version.
+  rpc Release(ReleaseRequest) returns (ReleaseResponse) {}
+
+  // Publish publishes a table descriptor change, writing the table
+  // descriptor to the KV store on success and notifying all watches
+  // of the new version. An error is returned if the version of the
+  // table descriptor is not equal to the current version + 1. An
+  // error is returned if there are two active versions of the table
+  // descriptor.
+  rpc Publish(PublishRequest) returns (PublishResponse) {}
+}
+```
+
+When a node starts up it will register a `Watch` with the lease master
+in order to receive a stream of version updates for all
+descriptors. Each node will maintain a map from `<descriptorID,
+version>` to `<TableDescriptor, expiration, localRefCount>`. When the
+node needs to use a descriptor for an operation it will look in its
+local map. If the descriptor is not there it will be read from the KV
+store and a lease will be acquired by sending an `Acquire` message to
+the lease master. The node will maintain the invariant that a new
+reference to a descriptor will only be added to the most recent
+version. When the reference count for an older version falls to 0 the
```

Comment:
The local refcount would be incremented when an operation starts using a descriptor and decremented when the surrounding transaction commits/aborts.